packages/core/lib/logger.js

```diff
@@ -1,96 +1,56 @@
-        timestamp: timestampFormatter
-      })
-    ]
-  });
-} else if (xrayLogLevel) {
-  logger = new (winston.Logger)({
-    transports: [
-      new (winston.transports.Console)({
-        formatter: outputFormatter,
-        level: xrayLogLevel,
-        timestamp: timestampFormatter
-      })
-    ]
-  });
-} else {
-  logger = new (winston.Logger)({});
-}
+var logger = {
+  error: createLoggerForLevel('error'),
+  info: createLoggerForLevel('info'),
+  warn: createLoggerForLevel('warn'),
+  debug: createLoggerForLevel('debug'),
+  setLevel: (level) => logLevel = calculateLogLevel(level),
+  getLevel: () => validLogLevels[logLevel]
+};
 
 /* eslint-disable no-console */
-if (process.env.LAMBDA_TASK_ROOT) {
-  logger.error = function(string) { console.error(string); };
-  logger.info = function(string) { console.info(string); };
-  logger.warn = function(string) { console.warn(string); };
-  logger.debug = function(string) { console.debug(string); };
+function createLoggerForLevel(level) {
+  var loggerLevel = validLogLevels.indexOf(level);
+  var consoleMethod = console[level] || console.log || (() => {});
+  return (message, meta) => {
+    if (loggerLevel >= logLevel) {
+      consoleMethod(formatLogMessage(level, message, meta));
+    }
+  };
 }
 /* eslint-enable no-console */
 
-function timestampFormatter() {
-  return format(new Date(), 'YYYY-MM-DD HH:mm:ss.SSS Z');
-}
+function calculateLogLevel(level) {
+  if (level) {
+    var normalisedLevel = level.toLowerCase();
+    var index = validLogLevels.indexOf(normalisedLevel);
+    return index >= 0 ? index : validLogLevels.length - 1;
+  }
 
-function outputFormatter(options) {
-  return options.timestamp() +' [' + options.level.toUpperCase() + '] '+
-    (options.message !== undefined ? options.message : '') +
-    (options.meta && Object.keys(options.meta).length ? '\n\t'+ JSON.stringify(options.meta) : '' );
+  // Silently ignore invalid log levels, default to highest level
+  return validLogLevels.length - 1;
 }
 
-/**
- * Polyfill for Object.keys
- * @see: https://developer.mozilla.org/en-US/docs/JavaScript/Reference/Global_Objects/Object/keys
- */
-
-if (!Object.keys) {
-  Object.keys = (function() {
-    'use strict';
-    var hasOwnProperty = Object.prototype.hasOwnProperty,
-      hasDontEnumBug = !({ toString: null }).propertyIsEnumerable('toString'),
-      dontEnums = [
-        'toString',
-        'toLocaleString',
-        'valueOf',
-        'hasOwnProperty',
-        'isPrototypeOf',
-        'propertyIsEnumerable',
-        'constructor'
-      ],
-      dontEnumsLength = dontEnums.length;
-
-    return function(obj) {
-      if (typeof obj !== 'object' && (typeof obj !== 'function' || obj === null)) {
-        throw new TypeError('Object.keys called on non-object');
-      }
+function createTimestamp() {
+  return format(new Date(), 'YYYY-MM-DD HH:mm:ss.SSS Z');
+}
 
-      var result = [], prop, i;
+function formatLogMessage(level, message, meta) {
```

Comment:
The fact that the time stamp is repeated for lambda did occur to me when I was writing this but won’t it be repeated for other things that write to cloudwatch as well? Say an ECS container for example?

At the time of writing I think I decided it was better to have a consistent log format that is produced regardless of the environment rather than trying to customise it for each different environment.

I’m happy to change it if that is the recommendation but, as I said, I don’t think this will avoid repeated timestamps in every situation.

An alternative could be to allow the user of the library to control the message format (either in total or just the inclusion of the time stamp) via another environment variable or api. I’m happy to prototype something along those lines if you want.

What do you think is the best approach?

Reply:
I think repeating the timestamp in CloudWatch is acceptable since CW treats their timestamp as a separate entity from the logging message, whereas Lambda actually interpolates the timestamp and log level into the logging message. I believe that is why the original author of the SDK special-cased logging in Lambda but not other environments.

If a user really cares about the logging format, they can always use the `setLogger()` API with their own logging object.

Reply:
Changed so that if the AWS_LAMBDA_TASK environment variable is set then the timestamp and logging level is not included in the outputted log line.